src/Shell/Task/ModelTask.php

```diff
@@ -344,25 +344,56 @@ public function findHasMany($model, array $associations)
                 }
             }
 
-            foreach ($otherSchema->columns() as $fieldName) {
-                $assoc = false;
-                if (!in_array($fieldName, $primaryKey) && $fieldName === $foreignKey) {
-                    $assoc = [
-                        'alias' => $otherModel->alias(),
-                        'foreignKey' => $fieldName
-                    ];
-                } elseif ($otherTableName === $tableName && $fieldName === 'parent_id') {
-                    $className = ($this->plugin) ? $this->plugin . '.' . $model->alias() : $model->alias();
-                    $assoc = [
-                        'alias' => 'Child' . $model->alias(),
-                        'className' => $className,
-                        'foreignKey' => $fieldName
-                    ];
+            $targetForeignKey = $primaryKey;
+            foreach ($targetForeignKey as $i => $fieldName) {
+                if ($fieldName === 'id') {
+                    $targetForeignKey[$i] = Inflector::singularize($model->table()) . '_id';
                 }
-                if ($assoc && $this->plugin && empty($assoc['className'])) {
-                    $assoc['className'] = $this->plugin . '.' . $assoc['alias'];
+            }
+
+            $possibleForeignKeys = $otherSchema->columns();
+            foreach ($possibleForeignKeys as $i => $fieldName) {
+                if ($fieldName === 'parent_id') {
+                    continue;
+                }
+                if (strlen($fieldName) <= 3 || !preg_match('/^.*_id$/', $fieldName)) {
+                    unset($possibleForeignKeys[$i]);
                 }
-                if ($assoc) {
+            }
+
+            $assoc = false;
+            if ($tableName !== $otherTableName
+                && array_diff(array_values($targetForeignKey), array_values($possibleForeignKeys)) === []
+            ) {
+                $assoc = [
+                    'alias' => $otherModel->alias(),
+                    'foreignKey' => $possibleForeignKeys,
+                ];
+            }
+            if ($otherTableName === $tableName && $fieldName === 'parent_id') {
```

Comment:
Thanks, there is an issue here. This is the migration I am using to test self join + composite 
```php
<?php
use Migrations\AbstractMigration;

class CreateCategories extends AbstractMigration
{
    /**
     * Change Method.
     *
     * More information on this method is available here:
     * http://docs.phinx.org/en/latest/migrations.html#the-change-method
     * @return void
     */
    public function change()
    {
        $table = $this->table('categories', [
            'id' => false,
            'primary_key' => [
                'project_id',
                'id',
        ]]);
        $table->addColumn('project_id', 'integer', [
            'default' => null,
            'limit' => 11,
            'null' => false,
        ]);
        $table->addColumn('id', 'integer', [
            'autoIncrement' => true,
            'limit' => 11
        ]);
        $table->addColumn('parent_id', 'integer', [
            'default' => null,
            'limit' => 11,
            'null' => true,
        ]);
        $table->addColumn('created', 'datetime', [
            'default' => 'CURRENT_TIMESTAMP',
            'null' => false,
            'timezone' => true,
        ]);
        $table->addColumn('modified', 'datetime', [
            'default' => 'CURRENT_TIMESTAMP',
            'null' => false,
            'timezone' => true,
        ]);
        $table->addColumn('name', 'string', [
            'default' => null,
            'limit' => 50,
            'null' => false,
        ]);
        $table->addForeignKey(['project_id', 'parent_id'], 'categories', ['project_id', 'id'], [
            'delete' => 'CASCADE',
            'update'=> 'NO_ACTION',
        ]);
        $table->addIndex([
            'name',
        ], [
            'name' => 'CATEGORIES_NAME_INDEX',
            'unique' => true,
        ]);
        $table->create();
    }
}
```
It throws errors in bake thus you were right on spot. Will prep a fix.